src/db.rs

```diff
@@ -1,67 +1,111 @@
 use anyhow::{anyhow, bail, Context, Result};
-use std::fs::{self, File};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use std::cmp::Ordering;
+use std::fs::{self, File, OpenOptions};
 use std::io::{self, BufRead, BufReader, BufWriter};
 use std::path::{Path, PathBuf};
 
+pub use i32 as DBVersion;
+
 pub struct DB {
-    path: PathBuf,
-    dirs: Vec<Dir>,
+    data: DBData,
     modified: bool,
+    path: PathBuf,
+    path_old: Option<PathBuf>,
 }
 
 impl DB {
     pub fn open<P: AsRef<Path>>(path: P) -> Result<DB> {
-        let path = path.as_ref().to_path_buf();
-
-        let dirs = match File::open(&path) {
+        let data = match File::open(&path) {
             Ok(file) => {
                 let reader = BufReader::new(&file);
-                bincode::deserialize_from(reader)
-                    .with_context(|| anyhow!("could not deserialize database"))?
+                bincode::config()
+                    .limit(config::DB_MAX_SIZE)
+                    .deserialize_from(reader)
+                    .context("could not deserialize database")?
             }
             Err(err) => match err.kind() {
-                io::ErrorKind::NotFound => Vec::<Dir>::new(),
-                _ => return Err(err).with_context(|| anyhow!("could not open database file")),
+                io::ErrorKind::NotFound => DBData::default(),
+                _ => return Err(err).context("could not open database file"),
             },
         };
 
+        if data.version != config::DB_VERSION {
+            bail!("this database version ({}) is unsupported", data.version);
+        }
+
         Ok(DB {
-            path,
-            dirs,
+            data,
             modified: false,
+            path: path.as_ref().to_path_buf(),
+            path_old: None,
+        })
+    }
+
+    pub fn open_and_migrate<P1, P2>(path_old: P1, path: P2) -> Result<DB>
+    where
+        P1: AsRef<Path>,
+        P2: AsRef<Path>,
+    {
+        let file = File::open(&path_old).context("could not open old database file")?;
+        let reader = BufReader::new(&file);
+
+        let dirs = bincode::config()
+            .limit(config::DB_MAX_SIZE)
+            .deserialize_from(reader)
+            .context("could not deserialize old database")?;
+
+        let data = DBData {
+            version: config::DB_VERSION,
+            dirs,
+        };
+
+        Ok(DB {
+            data,
+            modified: true,
+            path: path.as_ref().to_path_buf(),
+            path_old: Some(path_old.as_ref().to_path_buf()),
         })
     }
 
     pub fn save(&mut self) -> Result<()> {
         if self.modified {
             let path_tmp = self.get_path_tmp();
 
-            let file_tmp = File::create(&path_tmp)
-                .with_context(|| anyhow!("could not open temporary database file"))?;
+            let file_tmp = OpenOptions::new()
+                .write(true)
+                // ensure that we are not overwriting an existing file_tmp
```

Comment:
```suggestion
```

If you really want this comment, it should be above the assignment and something along the lines of (reflowing necessary):

```rust
// Calling `OpenOptions::new()` with `.create_new(true)` ensures that we are not overwriting an existing temporary file; it will return an error when the file already exists.
```